src/main/java/com/google/sps/servlets/TaskServlet.java

```diff
@@ -29,9 +37,75 @@
 @WebServlet("/tasks")
 public class TaskServlet extends HttpServlet {
   @Override
+  // doGet method retrieves tasks from datastore and responds with the HTML for each task fetched
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    // Serve the GET request sent from home page to fetch all the tasks
-    return;
+    UserService userService = UserServiceFactory.getUserService();
+    boolean userLoggedIn = userService.isUserLoggedIn();
+    String userId = userLoggedIn ? userService.getCurrentUser().getUserId() : "null";
+
+    String zipcode = request.getParameter("zipcode");
+    String country = request.getParameter("country");
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    Query query = new Query("Task").addSort("timestamp", SortDirection.DESCENDING);
+
+    // Creates list of filters
+    List<Query.Filter> filters = new ArrayList<Query.Filter>();
+    filters.add(new Query.FilterPredicate("zipcode", Query.FilterOperator.EQUAL, zipcode));
+    filters.add(new Query.FilterPredicate("country", Query.FilterOperator.EQUAL, country));
+
+    // Applies a category filter, if any
+    if (request.getParameterMap().containsKey("category")) {
+      String category = request.getParameter("category");
+      filters.add(new Query.FilterPredicate("category", Query.FilterOperator.EQUAL, category));
+    }
+
+    // Applies filters to query
+    query.setFilter(new Query.CompositeFilter(Query.CompositeFilterOperator.AND, filters));
+
+    // limits results to the 20 most recent tasks
+    List<Entity> results = datastore.prepare(query).asList(FetchOptions.Builder.withLimit(20));
```

Comment:
If we limit results to the 20 most recent tasks, I think we'd better notify the users on the web page. So, I would recommend adding a piece of information saying that "We limit to 20 most recent tasks" to `index.jsp`.

Reply:
Good idea! I'll update this now.